<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Commands &mdash; Chewbacca 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="Chewbacca 0.1 documentation" href="../index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="../index.html">Chewbacca 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="commands">
<h1>Commands<a class="headerlink" href="#commands" title="Permalink to this headline">¶</a></h1>
<div class="section" id="error-correction">
<span id="preclean"></span><h2>Error Correction<a class="headerlink" href="#error-correction" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="preclean.Preclean_Command.Preclean_Command">
<em class="property">class </em><code class="descclassname">preclean.Preclean_Command.</code><code class="descname">Preclean_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/preclean/Preclean_Command.html#Preclean_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#preclean.Preclean_Command.Preclean_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Attempts to fix minor sequencing errors caused by pyrosequencing.  By reducing errors prior to sequence assembly,
a greater number of paired reads can be sucessfully assembled.  Matching forward and reverse files should be         identically named,
except for a &lt;forward&gt;/&lt;reverse&gt; suffix that indicates the read orientation.  The two suffix conventions below         are supported. Choose ONE suffix style and stick to it!  Mixed suffixes are not supported.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">_forwards</span><span class="o">/</span><span class="n">_reverse</span>
<span class="ow">and</span>
<span class="n">_R1</span><span class="o">/</span><span class="n">_R2</span>
</pre></div>
</div>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>fastq file(s) with left reads .</li>
<li>fastq file(s) with right reads .</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>&lt;left reads file&gt;_corrected.fastq file(s).</li>
<li>&lt;right reads file&gt;_corrected.fastq file(s).</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<p>Assuming a forwards read file &#8216;Data_R1.fq&#8217; and a reverse reads file &#8216;Data_R1.fq&#8217;,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">./</span>
    <span class="n">Data_R1</span><span class="o">.</span><span class="n">fq</span>
    <span class="n">Data_R2</span><span class="o">.</span><span class="n">fq</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">preclean</span> <span class="pre">-f</span> <span class="pre">Data_R1.fq</span> <span class="pre">-r</span> <span class="pre">Data_R2.fq</span>&nbsp; <span class="pre">-o</span> <span class="pre">rslt</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
    <span class="n">Data_R1_corrected</span><span class="o">.</span><span class="n">fq</span>
    <span class="n">Data_R2_corrected</span><span class="o">.</span><span class="n">fq</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="preclean.Preclean_Command.Preclean_Command.default_program">
<code class="descname">default_program</code><a class="headerlink" href="#preclean.Preclean_Command.Preclean_Command.default_program" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal"><span class="pre">Preclean_Program_Bayeshammer</span></code></p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="assembling-sequences">
<span id="assemble"></span><h2>Assembling Sequences<a class="headerlink" href="#assembling-sequences" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="assemble.Assemble_Command.Assemble_Command">
<em class="property">class </em><code class="descclassname">assemble.Assemble_Command.</code><code class="descname">Assemble_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/assemble/Assemble_Command.html#Assemble_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#assemble.Assemble_Command.Assemble_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Assembles reads from two (forward and reverse) fastq files/directories.  For a set of k forward read files, and k
reverse read files, return k assembled files.  Matching forward and reverse files should be identically named,
except for a &lt;forward&gt;/&lt;reverse&gt; suffix that indicates the read orientation.  The two suffix conventions below         are supported.  Choose ONE suffix style and stick to it!  Mixed suffixes are not supported.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">_forwards</span><span class="o">/</span><span class="n">_reverse</span>
<span class="ow">and</span>
<span class="n">_R1</span><span class="o">/</span><span class="n">_R2</span>
</pre></div>
</div>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>fastq file(s) with left reads</li>
<li>fastq file(s) with right reads</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>fastq File(s) with assembled reads</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>Choose ONE suffix style and stick to it!  Mixed suffixes are not supported.                 e.g.                Sample_100_forwards.fq and Sample_100_reverse.fq will be assembled into Sample_100_assembled.fq.                 Simmilarly, Sample_100_R1.fq and Sample_100_R2.fq will be assembled into Sample_100_assembled.fq.                 However, Sample_100_forwards.fq and Sample_100_R2.fq are not guaranteed to be matched.</li>
<li>You can provide as many pairs of files as you wish as long as they follow exactly one of the above                  naming                 conventions.  If a &#8216;name&#8217; parameter is provided, it will be used as a filename (not path) prefix for                  all                 assembled sequence files.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong></p>
<p>Assuming a forwards read file &#8216;Data_R1.fq&#8217; and a reverse reads file &#8216;Data_R1.fq&#8217;,</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">./</span>
    <span class="n">Data_R1</span><span class="o">.</span><span class="n">fq</span>
    <span class="n">Data_R2</span><span class="o">.</span><span class="n">fq</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">assemble</span> <span class="pre">-n</span> <span class="pre">BALI</span> <span class="pre">-f</span> <span class="pre">Data_R1.fq</span>&nbsp; <span class="pre">-r</span> <span class="pre">Data_R2.fq</span>&nbsp; <span class="pre">-o</span> <span class="pre">rslt</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
    <span class="n">BALI_DATA</span><span class="o">.</span><span class="n">assembled</span><span class="o">.</span><span class="n">fq</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="assemble.Assemble_Command.Assemble_Command.default_program">
<code class="descname">default_program</code><a class="headerlink" href="#assemble.Assemble_Command.Assemble_Command.default_program" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal"><span class="pre">Assemble_Program_Pear</span></code></p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="demultiplexing-by-barcode">
<span id="demux"></span><h2>Demultiplexing by Barcode<a class="headerlink" href="#demultiplexing-by-barcode" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="demux.Demux_Command.Demux_Command">
<em class="property">class </em><code class="descclassname">demux.Demux_Command.</code><code class="descname">Demux_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/demux/Demux_Command.html#Demux_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#demux.Demux_Command.Demux_Command" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Given a set of files, each file is assigned a file_ID#.  Each file is then split into separate child files where</dt>
<dd>each file holds only sequences belonging to a single sample.  These child files are named using the sample name
for the sequences it lists, and the file_ID# of the file it came from.  Demuxing is based on the nucleotide
barcode prefixing each sequence.</dd>
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fasta/fastq files to demux.</li>
<li>A single .barcodes file: A <a class="reference internal" href="../file_types.html#barcodes"><span class="std std-ref">Barcodes file</span></a>.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>&lt;sample_name&gt;_&lt;file_id#&gt;_ demux.&lt;ext&gt; file(s) - &lt;fasta/fastq&gt; files, containing all the sequences from file             &lt;file_id#&gt;, which had a barcode corresponding to sample &lt;sample_name&gt;.</li>
<li>unmatched_&lt;file_id#&gt;_ demux.&lt;ext&gt; file(s) - &lt;fasta/fastq&gt; files, containing sequences from file             &lt;file_id#&gt;, whose barcode did not match any of those listed in the .barcodes file.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>The assignment of ID# to file should be treated as an arbitrary process and should not used for record             keeping.</li>
<li>Each input file will generate its own unmatched_* file (if applicable).</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">/</span>
    <span class="n">Data1</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
        <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1254</span>
        <span class="n">AGACGCAAAAAA</span>
        <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1255</span>
        <span class="n">AGTGTAAAAAAT</span>


    <span class="n">Data2</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
        <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1256</span>
        <span class="n">AGACGCAAAAAC</span>
        <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1257</span>
        <span class="n">AGTGTAAAAAAG</span>
        <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1258</span>
        <span class="n">CGTGTAAAAAAG</span>
<span class="o">./</span>
    <span class="n">Data</span><span class="o">.</span><span class="n">barcodes</span><span class="p">:</span>
        <span class="n">SampleA</span>        <span class="n">AGACGC</span>
        <span class="n">SampleB</span>        <span class="n">AGTGTA</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">demux_samples</span> <span class="pre">-i</span> <span class="pre">data/</span> <span class="pre">-b</span> <span class="pre">Data.barcodes</span> <span class="pre">-o</span> <span class="pre">rslt</span></code></p>
<p>Here, we see that Data1.fasta was assigned &#8216;0&#8217; as an ID#, while Data2.fasta was assigned &#8216;1&#8217; as an ID#.  Because     both files had sequences from SampleA, the sequences from Data1.fasta  were written to SampleA_0_demux.fastq,     and those sequences from Data2.fasta were written to SampleA_1_demux.fastq.  The same is true for SampleB.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
    <span class="n">SampleA_0_demux</span><span class="o">.</span><span class="n">fastq</span><span class="p">:</span>
        <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1254</span>
        <span class="n">AGACGCAAAAAA</span>

    <span class="n">SampleB_0_demux</span><span class="o">.</span><span class="n">fastq</span><span class="p">:</span>
        <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1255</span>
        <span class="n">AGTGTAAAAAAT</span>

    <span class="n">SampleA_1_demux</span><span class="o">.</span><span class="n">fastq</span><span class="p">:</span>
        <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1256</span>
        <span class="n">AGACGCAAAAAC</span>

    <span class="n">SampleB_1_demux</span><span class="o">.</span><span class="n">fastq</span><span class="p">:</span>
        <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1257</span>
        <span class="n">AGTGTAAAAAAG</span>

<span class="n">rslt_aux</span><span class="o">/</span>
    <span class="n">unmatched_0_demux</span><span class="o">.</span><span class="n">fastq</span><span class="p">:</span>
        <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1258</span>
        <span class="n">CGTGTAAAAAAG</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="demux.Demux_Command.Demux_Command.default_program">
<code class="descname">default_program</code><a class="headerlink" href="#demux.Demux_Command.Demux_Command.default_program" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal"><span class="pre">Demux_Program_Fastx</span></code></p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="sequence-renaming">
<span id="rename"></span><h2>Sequence Renaming<a class="headerlink" href="#sequence-renaming" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="rename.Rename_Command.Rename_Command">
<em class="property">class </em><code class="descclassname">rename.Rename_Command.</code><code class="descname">Rename_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/rename/Rename_Command.html#Rename_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#rename.Rename_Command.Rename_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Renames sequences in a file with their filename and a serial ID#.  Useful for simplifying complex naming systems     into     human-readable sequence names.  In order to ensure the correct sample names are preserved, it is reccomended that     this command be run immediately after the <span class="xref std std-ref">Demux Command</span>.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fasta/fastq files to rename.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>_renamed.&lt;ext&gt; file - A &lt;fasta/fastq&gt; file with the renamed sequences.</li>
<li>.samples file - A <a class="reference internal" href="../file_types.html#samples"><span class="std std-ref">Samples File</span></a>.</li>
<li>.mapping file - A <a class="reference internal" href="../file_types.html#mapping"><span class="std std-ref">Mapping file</span></a>.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last">
<li><p class="first">In order for the .samples file to correctly list the sample name of the sequences in a file, this command             should be run immediately after the Demux Command.</p>
</li>
<li><dl class="first docutils">
<dt>The <em>&#8211;clip</em> parameter tells Chewbacca that trailing _&lt;file_ID#&gt; (from the demultiplexing command) should not</dt>
<dd><p class="first last">be considered part of the sample name.  By default this is set to True, and should be fine.  If you notice
parts of your sample names getting clipped off in your .samples file, you should explicitly set this
parameter to False.</p>
</dd>
</dl>
</li>
<li><p class="first">Each input file will have a corresponding .samples, .mapping, and _renamed file.</p>
</li>
<li><p class="first">The .samples file is needed by downstream Chewbacca processes (Building the OTU Table).</p>
</li>
<li><p class="first">The .mapping file is purely for user convenience and record-keeping.</p>
</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">SampleA_0</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
    <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1256</span>
    <span class="n">AAAA</span>
    <span class="nd">@M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1257</span>
    <span class="n">AAAT</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">rename</span> <span class="pre">-i</span> <span class="pre">SampleA_0.fasta</span> <span class="pre">-o</span> <span class="pre">rslt</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span><span class="n">SampleA_0_renamed</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
    <span class="nd">@SampleA_ID0</span>
    <span class="n">AAAA</span>
    <span class="nd">@SampleA_ID1</span>
    <span class="n">AAAT</span>

<span class="n">rslt_samples</span><span class="o">/</span><span class="n">SampleA_0_renamed</span><span class="o">.</span><span class="n">samples</span><span class="p">:</span>
    <span class="n">SampleA_ID0</span> <span class="n">SampleA</span>
    <span class="n">SampleA_ID1</span> <span class="n">SampleA</span>

<span class="n">rslt_aux</span><span class="o">/</span><span class="n">SampleA_0_renamed</span><span class="o">.</span><span class="n">mapping</span><span class="p">:</span>
    <span class="n">M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1256</span> <span class="n">SampleA_ID0</span>
    <span class="n">M03292</span><span class="p">:</span><span class="mi">26</span><span class="p">:</span><span class="mi">000000000</span><span class="o">-</span><span class="n">AH6AG</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1101</span><span class="p">:</span><span class="mi">22127</span><span class="p">:</span><span class="mi">1257</span> <span class="n">SampleA_ID1</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="adapter-removal">
<span id="trim"></span><h2>Adapter Removal<a class="headerlink" href="#adapter-removal" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="clean.Clean_Adapters_Command.Clean_Adapters_Command">
<em class="property">class </em><code class="descclassname">clean.Clean_Adapters_Command.</code><code class="descname">Clean_Adapters_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/clean/Clean_Adapters_Command.html#Clean_Adapters_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#clean.Clean_Adapters_Command.Clean_Adapters_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes sequencing adapters (and preceeding barcodes) from sequences in input file(s).  Sequences should be in
the following format:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">BARCODE</span><span class="o">&gt;&lt;</span><span class="n">ADAPTER</span><span class="o">&gt;&lt;</span><span class="n">SEQUENCE</span><span class="o">&gt;&lt;</span><span class="n">RC_ADAPTER</span><span class="o">&gt;.</span>
</pre></div>
</div>
<p>Valid ADAPTER sequences, and their
reverse-complements (RC_ADAPTER) should be defined separately in a pair of fasta-formatted files.  Sequences
passed to this command should have already been demultiplexed, as this process will remove the identifying
barcode sequences.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fasta/fastq files to clean.</li>
<li>A single <a class="reference internal" href="../file_types.html#adapters"><span class="std std-ref">Adapters file</span></a> file</li>
<li>A single <a class="reference internal" href="../file_types.html#adaptersrc"><span class="std std-ref">RC Adapters file</span></a> file</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>&lt;filename&gt;_debarcoded.&lt;ext&gt; file(s) - &lt;fasta/fastq&gt; files, containing sequences with their leading                 adapters, trailing adapters, and barcodes removed.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>Be aware of the program-specific details around &#8216;N&#8217; nucleotide characters.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<p>Given Data_ID#1 with  barcode=AGACGC:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">./</span>
    <span class="n">Data</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
        <span class="nd">@Data_ID</span><span class="c1">#1</span>
        <span class="n">AGACGCGGWACWGGWTGAACWGTWTAYCCYCCATCGATCGATCGTGRTTYTTYGGNCAYCCNGARGTNTA</span>

    <span class="n">Data</span><span class="o">.</span><span class="n">adapters</span><span class="p">:</span>
        <span class="o">&gt;</span><span class="mi">1</span>
        <span class="n">GGWACWGGWTGAACWGTWTAYCCYCC</span>

    <span class="n">Data</span><span class="o">.</span><span class="n">adaptersRC</span><span class="p">:</span>
        <span class="o">&gt;</span><span class="n">first</span>
        <span class="n">TGRTTYTTYGGNCAYCCNGARGTNTA</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">trim_adapters</span>&nbsp; <span class="pre">-i</span> <span class="pre">Data.fasta</span> <span class="pre">-o</span> <span class="pre">rslt</span> <span class="pre">-a</span> <span class="pre">Data.adapters</span> <span class="pre">-arc</span> <span class="pre">Data.adapters_RC</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
    <span class="n">Data_debarcoded</span><span class="o">.</span><span class="n">fastq</span><span class="p">:</span>
        <span class="nd">@Data_ID</span><span class="c1">#1</span>
        <span class="n">ATCGATCGATCG</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="clean.Clean_Adapters_Command.Clean_Adapters_Command.default_program">
<code class="descname">default_program</code><a class="headerlink" href="#clean.Clean_Adapters_Command.Clean_Adapters_Command.default_program" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal"><span class="pre">Clean_Adapters_Program_Flexbar</span></code></p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="quality-cleaning">
<span id="trimmomatic"></span><h2>Quality Cleaning<a class="headerlink" href="#quality-cleaning" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="clean.Clean_Quality_Command.Clean_Quality_Command">
<em class="property">class </em><code class="descclassname">clean.Clean_Quality_Command.</code><code class="descname">Clean_Quality_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/clean/Clean_Quality_Command.html#Clean_Quality_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#clean.Clean_Quality_Command.Clean_Quality_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes regions of low quality from fastq-formatted reads.  These regions are likely sources of error, and would
be detrimental to other analytical process.  Input sequences to this command should have already been
demultiplexed, and had their barcodes/adapters removed.  Otherwise, the partial removal of these markers would
leave behind invalid fragments that would be difficult to detect and likely cause errors down-stream.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fastq files to clean.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>&lt;filename&gt;_cleaned.fastq file(s) - Fastq files, containing sequences with areas of low quality removed.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>Be aware of the program-specific details around &#8216;N&#8217; nucleotide characters.</li>
<li>Be aware of the program-specific defaults for minimum surviving sequence lengths.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>./
    Data.fasta:
        @Data_ID#1
        GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTCTTTACAG
        +
        !zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz%%%zzzz
</pre></div>
</div>
<p>The command below asks Chewbacca to trim away any section of length 3 NT in Data_ID#1 that has quality worse             than 20, keeping the             longer of the remaining ends.  If the remaining sequence at the end of this process is shorter than 15 NT,             discard the whole sequence (these values are chosen for illustrative purposes).</p>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">clean_seqs</span> <span class="pre">-i</span> <span class="pre">Data.fasta</span> <span class="pre">-o</span> <span class="pre">rslt</span> <span class="pre">-m</span> <span class="pre">15</span> <span class="pre">-w</span> <span class="pre">3</span> <span class="pre">-q</span> <span class="pre">20</span></code></p>
<p>Note that the &#8216;TTT&#8217; subsequence has been cut, because its average quality (5) is less than the threshold (20).         After this cut, the longest surviving subsequence (the subsequence to the left of the cut) was kept, and the         shorter subsequence (to the right of the cut) was discarded.  Because the final sequence is longer than 15NT,         it is kept and written to the output file.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>rslt/
    Data_cleaned.fastq:
        @Data_ID#1
        GATTTGGGGTTCAAAGCAGTATCGATCAAATAGTAAATCCATTTGTTCAACTC
        +
        !zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz
</pre></div>
</div>
<dl class="attribute">
<dt id="clean.Clean_Quality_Command.Clean_Quality_Command.default_program">
<code class="descname">default_program</code><a class="headerlink" href="#clean.Clean_Quality_Command.Clean_Quality_Command.default_program" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal"><span class="pre">Clean_Quality_Program_Trimmomatic</span></code></p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="file-conversion">
<span id="to-fasta"></span><h2>File Conversion<a class="headerlink" href="#file-conversion" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="util.Convert_Fastq_Fasta_Command.Convert_Fastq_Fasta_Command">
<em class="property">class </em><code class="descclassname">util.Convert_Fastq_Fasta_Command.</code><code class="descname">Convert_Fastq_Fasta_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/util/Convert_Fastq_Fasta_Command.html#Convert_Fastq_Fasta_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#util.Convert_Fastq_Fasta_Command.Convert_Fastq_Fasta_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a Fasta-formatted file to a FastQ-formatted file.  Useful for reducing data size and preparing for
fasta-only operations.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fastq files to convert to fasta.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>&lt;filename&gt;.fasta file(s) - Converted fasta files.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">./</span>
    <span class="n">Data</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
        <span class="nd">@Data_ID</span><span class="c1">#1</span>
        <span class="n">AGACGCGGWACWGGWTGAACWGTWTAYCCYCCATCGATCGATCGTGRTTYTTYGGNCAYCCNGARGTNTA</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">trim_adapters</span>&nbsp; <span class="pre">-i</span> <span class="pre">Data.fasta</span> <span class="pre">-o</span> <span class="pre">rslt</span> <span class="pre">-a</span> <span class="pre">Data.adapters</span> <span class="pre">-arc</span> <span class="pre">Data.adapters_RC</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
    <span class="n">Data_debarcoded</span><span class="o">.</span><span class="n">fastq</span><span class="p">:</span>
        <span class="nd">@Data_ID</span><span class="c1">#1</span>
        <span class="n">ATCGATCGATCG</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="dereplication">
<span id="derep"></span><h2>Dereplication<a class="headerlink" href="#dereplication" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="dereplicate.Dereplicate_Command.Dereplicate_Command">
<em class="property">class </em><code class="descclassname">dereplicate.Dereplicate_Command.</code><code class="descname">Dereplicate_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/dereplicate/Dereplicate_Command.html#Dereplicate_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dereplicate.Dereplicate_Command.Dereplicate_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Dereplicates a fasta file by grouping identical reads together under one representative sequence.  The
number of duplicate/replicant sequences each representative sequence represents is given by a &#8216;replication count&#8217; at
the end of
the sequence name in output fasta file.  If a .groups file is provided, then previous replication counts will
be take in into account (e.g. Imagine a representative sequence X that represents 3 sequences.  If X is found to be
a replicant of another sequence Y, X will add 3 to replication count of Y).  Replication counts are denoted with a
suffix of &#8216;_K&#8217; on the sequence name, where K is the replication count for the group that sequence represents.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fasta files to dereplicate.</li>
<li>Optional: <a class="reference internal" href="../file_types.html#groups"><span class="std std-ref">Groups File</span></a> - A list of representative names and the names of their replicant                                             sequences.  You likely have one of these files if you&#8217;ve previously run a                                             clustering or dereplication command.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>_counts.fasta file - A fasta file with unique sequences and their replication counts.</li>
<li>_derep:ref:<cite>.groups</cite> - A list of representative names and the names of their replicant                                             sequences.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>This command only dereplicates within each fasta file (not across all files).             This means a sequence in one file will be unique within that file, but might exist in another file.             To ensure sequences are uniqe across an entire dataset, merge all fasta files into one file, then             dereplicate that fasta file.</li>
<li>Each input file will generate a corresponding _count file.</li>
<li>If an input .groups file is not provided, then each input fasta file will generate a new groups file named             &lt;file_name&gt;_derep.groups.  If an input .groups file IS provided, then a single groups file named             &#8216;dereplicated_updated.groups&#8217; will be generated.</li>
<li>The output .groups file is needed by downstream Chewbacca processes (Dereplication, Clustering,                Building the OTU Table).</li>
<li>The order of sequence names in the *_counts.fasta and .groups file is arbitrary.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">./</span>
    <span class="n">Data</span><span class="o">.</span><span class="n">fasta</span>
        <span class="o">&gt;</span><span class="n">seq1</span>
        <span class="n">AAA</span>
        <span class="o">&gt;</span><span class="n">seq2</span>
        <span class="n">AAA</span>
        <span class="o">&gt;</span><span class="n">seq3</span>
        <span class="n">AAAG</span>
        <span class="o">&gt;</span><span class="n">seq4</span>
        <span class="n">AAAGT</span>
        <span class="o">&gt;</span><span class="n">seq7</span>
        <span class="n">AAAGT</span>

    <span class="n">test</span><span class="o">.</span><span class="n">groups</span>
        <span class="n">seq4</span>    <span class="n">seq4</span> <span class="n">seq5</span> <span class="n">seq6</span>
</pre></div>
</div>
<p>In the above example, test.groups indicates that seq4 is a sequence that has previously been identified as a
representative (in some earlier round of clustering or dereplication).  Note that seq4 is a representative
for a &#8216;group&#8217; of identical sequences and therefore listed within that group.</p>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">dereplicate_fasta</span> <span class="pre">-i</span> <span class="pre">Data.fasta</span> <span class="pre">-o</span> <span class="pre">rslt</span> <span class="pre">-g</span> <span class="pre">test.groups</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span><span class="n">Data_counts</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
    <span class="o">&gt;</span><span class="n">seq4_4</span>
    <span class="n">AAAGT</span>
    <span class="o">&gt;</span><span class="n">seq1_2</span>
    <span class="n">AAA</span>
    <span class="o">&gt;</span><span class="n">seq3_1</span>
    <span class="n">AAAG</span>

<span class="n">rslt_groups_files</span><span class="o">/*.</span><span class="n">groups</span><span class="p">:</span>
    <span class="n">seq3</span>        <span class="n">seq3</span>
    <span class="n">seq1</span>        <span class="n">seq2</span> <span class="n">seq1</span>
    <span class="n">seq4</span>        <span class="n">seq7</span> <span class="n">seq6</span> <span class="n">seq5</span> <span class="n">seq4</span>
</pre></div>
</div>
<p>Notice that Data_counts.fasta lists the unique sequences from Data.fasta, and their replication counts.  Also
notice that seq4 had previous replication data (stored in the Data.groups file).</p>
<dl class="attribute">
<dt id="dereplicate.Dereplicate_Command.Dereplicate_Command.default_program">
<code class="descname">default_program</code><a class="headerlink" href="#dereplicate.Dereplicate_Command.Dereplicate_Command.default_program" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal"><span class="pre">Dereplicate_Program_Vsearch</span></code></p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="file-splitting">
<span id="partition"></span><h2>File Splitting<a class="headerlink" href="#file-splitting" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="util.Partition_Command.Partition_Command">
<em class="property">class </em><code class="descclassname">util.Partition_Command.</code><code class="descname">Partition_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/util/Partition_Command.html#Partition_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#util.Partition_Command.Partition_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>A utility command that partitions a fasta/fastq file into  a set of files (of the same file format), with a         user-specified (maximum) number of sequences per file.  Allows users to partition a large file into segments,         and perform discrete operations in parallel over those segments.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fasta/fastq files to partition.</li>
<li>C: An integer defining the maximum number of sequences per file</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>&lt;filename&gt;_part_&lt;part_#&gt;.&lt;ext&gt; file(s) - &lt;fasta/fastq&gt; files, with at most C sequences per file.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>./
    Data.fq:
        @Data_ID1
        GATTTGGGG
        +
        !zzzzzzzzz
        @Data_ID2
        GATTTGGGG
        +
        !zzzzzzzzz
        @Data_ID3
        GATTTGGGG
        +
        !zzzzzzzzz
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">convert_fastq_to_fasta</span> <span class="pre">-i</span> <span class="pre">Data.fq</span> <span class="pre">-o</span> <span class="pre">rslt/</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
    <span class="n">Data</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
        <span class="nd">@Data_ID1</span>
        <span class="n">GATTTGGGG</span>
        <span class="nd">@Data_ID2</span>
        <span class="n">GATTTGGGG</span>
        <span class="nd">@Data_ID3</span>
        <span class="n">GATTTGGGG</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="file-merging">
<span id="merge"></span><h2>File Merging<a class="headerlink" href="#file-merging" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="util.Merge_Command.Merge_Command">
<em class="property">class </em><code class="descclassname">util.Merge_Command.</code><code class="descname">Merge_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/util/Merge_Command.html#Merge_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#util.Merge_Command.Merge_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Concatenates multiple files into a single file.  Useful for combining the results of a parallel operation, or         when preparing for cross-sample derepication.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>A set of files to merge.</li>
<li>An &lt;output_filename&gt;.</li>
<li>An &lt;output_prefix&gt;.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>&lt;output_filename&gt;.&lt;output_prefix&gt; - A file consisting of all the input files concatenated together.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>The order of the content in the concatenated files is not guaranteed.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>targets/
    Data.fq:
        @Data_ID1
        GATTTGGGG
        +
        !zzzzzzzzz

    Data2.fa:
        @Data_ID1
        GATTTGGGG

    Blah.txt
        Hello World!
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">merge_files</span> <span class="pre">-i</span> <span class="pre">targets/</span> <span class="pre">-o</span> <span class="pre">rslt/</span> <span class="pre">-f</span> <span class="pre">txt</span> <span class="pre">-n</span> <span class="pre">Merged</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span>rslt/
    Merged.txt:
        Hello World!
        @Data_ID1
        GATTTGGGG
        +
        !zzzzzzzzz
        @Data_ID1
        GATTTGGGG
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="file-cleaning">
<span id="ungap"></span><h2>File Cleaning<a class="headerlink" href="#file-cleaning" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="util.Ungap_Command.Ungap_Command">
<em class="property">class </em><code class="descclassname">util.Ungap_Command.</code><code class="descname">Ungap_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/util/Ungap_Command.html#Ungap_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#util.Ungap_Command.Ungap_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Removes target characters from a fasta/fastq file.  Useful for removing gap characters from sequence alignments.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fasta/fastq files to clean.</li>
<li>A string of gap characters to remove.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>*_cleaned.&lt;ext&gt; file - A &lt;fasta/fastq&gt; file with gap characters removed from its sequences.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Data</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
    <span class="o">&gt;</span><span class="n">seq1</span>
    <span class="n">AAAAA</span><span class="o">.</span><span class="n">A</span><span class="o">*</span><span class="n">A</span><span class="o">-</span><span class="n">A</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">ungap_fasta</span> <span class="pre">-i</span> <span class="pre">Data.fasta</span> <span class="pre">-o</span> <span class="pre">rslt</span> <span class="pre">-f</span> <span class="pre">fasta</span> <span class="pre">-g</span> <span class="pre">&quot;.*-&quot;</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span><span class="n">Data</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
    <span class="o">&gt;</span><span class="n">seq1</span>
    <span class="n">AAAAAAAA</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="deep-cleaning">
<span id="macse"></span><h2>Deep Cleaning<a class="headerlink" href="#deep-cleaning" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="clean.Clean_Deep_Command.Clean_Deep_Command">
<em class="property">class </em><code class="descclassname">clean.Clean_Deep_Command.</code><code class="descname">Clean_Deep_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/clean/Clean_Deep_Command.html#Clean_Deep_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#clean.Clean_Deep_Command.Clean_Deep_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs an intensive deep-cleaning of sequences to eliminate frameshifts, detect chimeras,
and determine sequence orientation.  Input files to this command should first be dereplicated.  Doing so will
reduce the total number of alignments required, and reduce computation time.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fasta/fastq files to deep clean (nucleotide sequences).</li>
<li>One reference fasta (nucleotide sequences).</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>*_AA - Amino Acid Alignment file, including reference sequences.</li>
<li>*_log.csv - A log listing each input sequence, and deep cleaning results for each sequence.</li>
<li>*_NT -  Nucleotide Alignment file, including reference sequences.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>Sequences that do not meet quality cleaning standards are dropped.</li>
<li>The output files contain reference sequences, and odd alignment characters.  Both of these need to be             removed by running the Clean_Deep_Repair Command.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">Data</span><span class="o">.</span><span class="n">fasta</span>
<span class="n">BIOCODE</span><span class="o">.</span><span class="n">fa</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">macseAlign</span> <span class="pre">-i</span> <span class="pre">Data.fasta</span> <span class="pre">-o</span> <span class="pre">rslt</span> <span class="pre">-d</span> <span class="pre">BIOCODE.fa</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span><span class="n">Data_AA</span>
<span class="n">rslt</span><span class="o">/</span><span class="n">Data_NT</span>
<span class="n">rslt</span><span class="o">/</span><span class="n">Data_log</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="clean.Clean_Deep_Command.Clean_Deep_Command.default_program">
<code class="descname">default_program</code><a class="headerlink" href="#clean.Clean_Deep_Command.Clean_Deep_Command.default_program" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal"><span class="pre">Clean_Deep_Program_Macse</span></code></p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="deep-cleaning-repair">
<span id="macse-clean"></span><h2>Deep Cleaning Repair<a class="headerlink" href="#deep-cleaning-repair" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="clean.Clean_Deep_Repair_Command.Clean_Deep_Repair_Command">
<em class="property">class </em><code class="descclassname">clean.Clean_Deep_Repair_Command.</code><code class="descname">Clean_Deep_Repair_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/clean/Clean_Deep_Repair_Command.html#Clean_Deep_Repair_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#clean.Clean_Deep_Repair_Command.Clean_Deep_Repair_Command" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Cleans aligned files by removing gap characters and reference sequences from the file.  Sequences passed to this</dt>
<dd>command should have previously been aligned.</dd>
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>*_AA - Amino Acid Alignment file, including reference sequences.</li>
<li>*_log.csv - A log listing each input sequence, and deep cleaning results for each sequence.</li>
<li>*_NT -  Nucleotide Alignment file, including reference sequences.</li>
<li>Nucleotide reference fasta.</li>
<li>* The original fasta files that were passed in to the Clean_Deep Command</li>
<li>* The Nucleotide reference fasta that was passed to the Clean_Deep Command</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>*_MERGED.fasta - A clean fasta file with all the surviving sequences from deep cleaning.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>A single *_MERGED.fasta is generated regardless of the number of input files.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">BIOCODE</span><span class="o">.</span><span class="n">fa</span>

<span class="n">originalData</span><span class="o">/</span><span class="n">Data</span><span class="o">.</span><span class="n">fasta</span>

<span class="nb">input</span><span class="o">/</span>
    <span class="n">Data_AA</span>
    <span class="n">Data_NT</span>
    <span class="n">Data_log</span><span class="o">.</span><span class="n">csv</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">-i</span> <span class="pre">input/</span> <span class="pre">-o</span> <span class="pre">out/</span> <span class="pre">-d</span>&nbsp; <span class="pre">BIOCODE.fa</span> <span class="pre">-s</span> <span class="pre">originalData/</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">out</span><span class="o">/</span>
    <span class="n">MACSE_OUT_MERGED</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="clean.Clean_Deep_Repair_Command.Clean_Deep_Repair_Command.default_program">
<code class="descname">default_program</code><a class="headerlink" href="#clean.Clean_Deep_Repair_Command.Clean_Deep_Repair_Command.default_program" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <code class="xref py py-class docutils literal"><span class="pre">Clean_Deep_Repair_Program_Macse</span></code></p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="sequence-clustering">
<span id="cluster"></span><h2>Sequence Clustering<a class="headerlink" href="#sequence-clustering" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="cluster.Cluster_Command.Cluster_Command">
<em class="property">class </em><code class="descclassname">cluster.Cluster_Command.</code><code class="descname">Cluster_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/cluster/Cluster_Command.html#Cluster_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cluster.Cluster_Command.Cluster_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Clusters a set of fasta files.  This command generates a fasta file of unique sequences
(each representing a cluster) and a .groups file.  This command also takes an optional .groups file containing
replication data from previous commands.  If a .groups file is supplied, only one output .groups file is generated
(regardless of the number of inputs).</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fasta files to cluster.</li>
<li>Optional: <a class="reference internal" href="../file_types.html#groups"><span class="std std-ref">Groups File</span></a> - A list of representative names and the names of their replicant                                             sequences.  You likely have one of these files if you&#8217;ve previously run a                                             clustering or dereplication command.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>*.fasta file - A fasta file with unique sequences and their replication counts.</li>
<li>*.groups  - A <a class="reference internal" href="../file_types.html#groups"><span class="std std-ref">Groups File</span></a></li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last">
<li><p class="first">The input fasta file(s) should have been dereplicated before clustering.</p>
</li>
<li><dl class="first docutils">
<dt>For a single experiment with multiple fasta files, it is best to merge all input fasta files, dereplicate</dt>
<dd><p class="first last">them, then cluster the single merged and dereplicated fasta file.  This provides the best OTU groupings.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">./</span>
    <span class="n">Data</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
        <span class="o">&gt;</span><span class="n">seq1_3</span>
        <span class="n">AAAAAAAAAA</span>
        <span class="o">&gt;</span><span class="n">seq2_1</span>
        <span class="n">ATAAAAAAAA</span>
        <span class="o">&gt;</span><span class="n">seq3_1</span>
        <span class="n">TTTTTTTTTT</span>
        <span class="o">&gt;</span><span class="n">seq4_1</span>
        <span class="n">TTTTTTATTT</span>
        <span class="o">&gt;</span><span class="n">seq5_1</span>
        <span class="n">TTTTTTATCT</span>


    <span class="n">Data</span><span class="o">.</span><span class="n">groups</span><span class="p">:</span>
        <span class="n">seq1</span>    <span class="n">seq6</span> <span class="n">seq1</span> <span class="n">seq7</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">cluster_seqs</span> <span class="pre">-i</span> <span class="pre">Data.fasta</span> <span class="pre">-o</span> <span class="pre">rslt</span> <span class="pre">-g</span> <span class="pre">Data.groups</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
    <span class="n">Data_clustered_seeds</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
        <span class="o">&gt;</span><span class="n">seq1_4</span>
        <span class="n">AAAAAAAAAA</span>
        <span class="o">&gt;</span><span class="n">seq3_3</span>
        <span class="n">TTTTTTTTTT</span>

<span class="n">rslt_groups_files</span><span class="o">/</span>
    <span class="n">postcluster_updated</span><span class="o">.</span><span class="n">groups</span><span class="p">:</span>
        <span class="n">seq3</span>    <span class="n">seq3</span> <span class="n">seq5</span> <span class="n">seq4</span>
        <span class="n">seq1</span>    <span class="n">seq2</span> <span class="n">seq1</span> <span class="n">seq7</span> <span class="n">seq6</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="otu-table-construction">
<span id="build-otu"></span><h2>OTU Table Construction<a class="headerlink" href="#otu-table-construction" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="otu.Build_OTU_Table_Command.Build_OTU_Table_Command">
<em class="property">class </em><code class="descclassname">otu.Build_OTU_Table_Command.</code><code class="descname">Build_OTU_Table_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/otu/Build_OTU_Table_Command.html#Build_OTU_Table_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#otu.Build_OTU_Table_Command.Build_OTU_Table_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds an OTU table using a .groups, .samples, and .barcodes file.  The OTU table shows OTU (group) abundance by
sample.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more <a class="reference internal" href="../file_types.html#samples"><span class="std std-ref">Samples File</span></a>.</li>
<li>One or more <a class="reference internal" href="../file_types.html#barcodes"><span class="std std-ref">Barcodes file</span></a>.</li>
<li>one or more <a class="reference internal" href="../file_types.html#groups"><span class="std std-ref">Groups File</span></a>.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>matrix.txt - A tab-delimited table mapping OTUs (groups) to their abundance in each sample.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>A sequence name may not appear in more than one group file (or more than one line in a gropus file for             that matter!).</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">./</span>
    <span class="n">test</span><span class="o">.</span><span class="n">barcodes</span>
        <span class="n">Sample1</span> <span class="n">aaaaaa</span>
        <span class="n">Sample2</span> <span class="n">aaaaat</span>
        <span class="n">Sample3</span> <span class="n">aaaaac</span>
        <span class="n">Sample4</span> <span class="n">aaaaag</span>

    <span class="n">test</span><span class="o">.</span><span class="n">groups</span>
        <span class="n">seq3</span>    <span class="n">seq3</span> <span class="n">seq5</span> <span class="n">seq4</span>
        <span class="n">seq1</span>    <span class="n">seq2</span> <span class="n">seq1</span> <span class="n">seq7</span> <span class="n">seq6</span>

    <span class="n">test</span><span class="o">.</span><span class="n">samples</span>
        <span class="n">seq1</span>    <span class="n">Sample1</span>
        <span class="n">seq2</span>    <span class="n">Sample1</span>
        <span class="n">seq3</span>    <span class="n">Sample1</span>
        <span class="n">seq4</span>    <span class="n">Sample2</span>
        <span class="n">seq5</span>    <span class="n">Sample2</span>
        <span class="n">seq6</span>    <span class="n">Sample2</span>
        <span class="n">seq7</span>    <span class="n">Sample3</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">build_matrix</span> <span class="pre">-b</span> <span class="pre">test.barcodes</span> <span class="pre">-g</span>&nbsp; <span class="pre">test.groups</span> <span class="pre">-s</span> <span class="pre">test.samples</span> <span class="pre">-o</span> <span class="pre">rslt/</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
    <span class="n">matrix</span><span class="o">.</span><span class="n">txt</span>
        <span class="n">OTU</span>     <span class="n">Sample1</span> <span class="n">Sample2</span> <span class="n">Sample3</span> <span class="n">Sample4</span>
        <span class="n">seq3</span>    <span class="mi">1</span>       <span class="mi">2</span>       <span class="mi">0</span>       <span class="mi">0</span>
        <span class="n">seq1</span>    <span class="mi">2</span>       <span class="mi">1</span>       <span class="mi">1</span>       <span class="mi">0</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="otu-identification">
<span id="id-otu"></span><h2>OTU Identification<a class="headerlink" href="#otu-identification" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="otu.Query_OTU_DB_Command.Query_OTU_DB_Command">
<em class="property">class </em><code class="descclassname">otu.Query_OTU_DB_Command.</code><code class="descname">Query_OTU_DB_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/otu/Query_OTU_DB_Command.html#Query_OTU_DB_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#otu.Query_OTU_DB_Command.Query_OTU_DB_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Aligns sequences in a fasta file against those in a reference database in order to determine OTU identity.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fasta files containing sequences to identify.</li>
<li>A curated fasta file of high quality sequences and known species.</li>
<li>A database containing taxonomic identifiers for sequences in the curated fasta file.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>A <a class="reference internal" href="../file_types.html#tax"><span class="std std-ref">Tax file</span></a>.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>The files COI.fasta and ncbi.db are included in the Chewbacca Docker distributions.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">~/</span><span class="n">ARMS</span><span class="o">/</span><span class="n">refs</span><span class="o">/</span>

    <span class="n">COI</span><span class="o">.</span><span class="n">fasta</span> <span class="c1"># A precompiled fasta file of COI data from NCBI.</span>
        <span class="o">&gt;</span><span class="mi">94483305</span>
        <span class="n">AGGACGGATCAGACGAAGAGGGGCGTTTGGTATTGGGTTATGGCAGGGGGTTTTATATTGATAATTGTTGTGATGAAATT</span>
        <span class="n">GATGGCCCCTAAGATAGAGGAGACACCTGCTAGGTGTAAGGAGAAGATGGTTAGGTCTACGGAGGCTCCAGGGTGGGAGT</span>

    <span class="n">ncbi</span><span class="o">.</span><span class="n">db</span> <span class="c1"># A precompiled database of (Taxa) for the entries in &#39;COI.fasta&#39;.</span>


<span class="n">data</span><span class="o">/</span>
    <span class="n">Data</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
        <span class="o">&gt;</span><span class="n">seq1</span>
        <span class="n">GAATAGGTGTTGGTATAGAATGGGGTCTCCTCCTCCGGCGGGGTCGAAGAAGGTGGTGTTGAGGTTGCGGTCTGTTAGTAGTATAGTGATGCCAGCAG</span>
        <span class="n">CTAGGACTGGGAGAGATAGGAGAAGTAGGACTGCTGTGATTAGGACGGATCAGACGAAGAGGGGCGTTTGGTATTGGGTTATGGCAGGGGGTTTTATA</span>
        <span class="n">TTGATAATTGTTGTGAGGAAATTGATGGCCCCTAAGATAGAGGAGACACCTGCTAGGTGTAAGGAGAAGATGGTTAGGTCTACGGAGGCTCCAGGGTG</span>
        <span class="n">GGAGTAGTTCCCTGCTAA</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">query_db</span> <span class="pre">-i</span> <span class="pre">Data.fasta</span> <span class="pre">-o</span> <span class="pre">out/</span> <span class="pre">-r</span> <span class="pre">~/ARMS/refs/COI.fasta</span> <span class="pre">-d</span> <span class="pre">~/ARMS/refs/ncbi.db</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
    <span class="n">Data_result</span><span class="o">.</span><span class="n">out</span>
        <span class="n">seq1</span>    <span class="mi">94483305</span>        <span class="mf">99.4</span>    <span class="mi">173</span>     <span class="mf">55.4</span>    <span class="n">Chordata</span><span class="p">:</span><span class="n">Mammalia</span><span class="p">:</span><span class="n">Primates</span><span class="p">:</span><span class="n">Hominidae</span><span class="p">:</span><span class="n">Homo</span><span class="p">:</span><span class="n">Homo</span> <span class="n">sapiens</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="otu.Query_OTU_Fasta_Command.Query_OTU_Fasta_Command">
<em class="property">class </em><code class="descclassname">otu.Query_OTU_Fasta_Command.</code><code class="descname">Query_OTU_Fasta_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/otu/Query_OTU_Fasta_Command.html#Query_OTU_Fasta_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#otu.Query_OTU_Fasta_Command.Query_OTU_Fasta_Command" title="Permalink to this definition">¶</a></dt>
<dd><p>Aligns sequences in a fasta file against those in a reference fasta in order to determine OTU identity.</p>
<dl class="docutils">
<dt><strong>Inputs</strong>:</dt>
<dd><ul class="first last simple">
<li>One or more fasta files containing sequences to identify.</li>
<li>A curated fasta file of high quality sequences and known species.</li>
<li>A two-column, tab-delimited text file mapping sequence names in the curated fasta file to taxonomic             identifiers.</li>
</ul>
</dd>
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>A <a class="reference internal" href="../file_types.html#tax"><span class="std std-ref">Tax file</span></a>.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>The files &#8216;bold.fna&#8217; and &#8216;seq_lin.mapping&#8217; are included in the Chewbacca Docker distributions.</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">~/</span><span class="n">ARMS</span><span class="o">/</span><span class="n">data</span><span class="o">/</span>
    <span class="n">bold</span><span class="o">.</span><span class="n">fna</span> <span class="c1"># A precompiled fasta file of data from BOLD.</span>
        <span class="o">&gt;</span><span class="n">GBMAA1117</span><span class="o">-</span><span class="mi">14</span>
        <span class="n">GGGCTTTTGCGGGTATGATAGGAACAGCATTTAGTATGCTTATTAGGTTAGAACTATCTTCCCCAGGGTCTATGTTAGGAGATGATCATTTATATAAT</span>
        <span class="n">GTTATAGTAACAGCTCATGCATTTGTAATGATATTTTTTTTAGTTATGCCAGTAATGATTGGGGGTTTTGGTAATTGGTTAGTACCTTTATATATTGG</span>
        <span class="n">TGCCCCGGATATGGCTTTTCCTAGATTAAATAATATTAGTTTTTGGTTATTACCTCCGGCGCTTACTTTATTATTAGGTTCGGCTTTTGTAGAACAAG</span>
        <span class="n">GGGCTGGGACAGGTTGGACAGTTTATCCGCCTTTATTTAGTATTCAAACTCATTCTGGGGGGTCTGTGGATATGGTAATATTTAGTTTACATTTAGCT</span>
        <span class="n">GGAATATCTTCTATATTAGGGGCTATGAATTTTATAACAACAATCTTTAATATGAGGTCTCCGGGAGTAACTATGGATAGAATGCCTTTATTTGTTTG</span>
        <span class="n">ATCTGTTTTAGTAACTGCTTTTTTATTATTATTATCATTGCCAGTATTAGCTGGTGCCATAACAAGTCTTTTAACCGATCGAGATTTTAATACTACAT</span>
        <span class="n">TT</span>

    <span class="n">seq_lin</span><span class="o">.</span><span class="n">mapping</span> <span class="c1"># A precompiled two-column tab file of (Taxa) for the entries in &#39;bold.fna&#39;.</span>
        <span class="n">GBMAA1117</span><span class="o">-</span><span class="mi">14</span>    <span class="n">Animalia</span><span class="p">;</span><span class="n">Porifera</span><span class="p">;</span><span class="n">Demospongiae</span><span class="p">;</span><span class="n">Haplosclerida</span><span class="p">;</span><span class="n">Phloeodictyidae</span><span class="p">;;</span><span class="n">Calyx</span><span class="p">;</span><span class="n">Calyx</span> <span class="n">podatypa</span>

<span class="o">./</span>
    <span class="n">Data</span><span class="o">.</span><span class="n">fasta</span><span class="p">:</span>
        <span class="o">&gt;</span><span class="n">seq1</span>
        <span class="n">ACTATCAGGCATTCAAGCCCATTCAGGGGGAGCAGTAGATATGGCTATATTTAGTCTACATCTAGCTGGTGTATCCTCTATTTTAAGTTCTATAAACT</span>
        <span class="n">TTATAACTACTATAATTAATATGAGGGTTCCTGGGATGAGTATGCATAGATTACCTCTATTCGTATGGTCTGTATTAGTTACTACAATATTATTGTTG</span>
        <span class="n">TTATCTTTACCAGTATTAGCTGGTGGAATTACAATGTTATTGACAGATAGAAATTTTAATACAACATTCTTTGACCCTGCGGGAGGAGGAGATCCTAT</span>
        <span class="n">TTTATTCCAGCACTTATTT</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">query_fasta</span> <span class="pre">-i</span> <span class="pre">Data.fasta</span> <span class="pre">-o</span> <span class="pre">rslt</span> <span class="pre">-r</span> <span class="pre">~/ARMS/data/bold.fna</span> <span class="pre">-x</span> <span class="pre">~/ARMS/data/seq_lin.mapping</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
        <span class="n">Data_result</span><span class="o">.</span><span class="n">out</span>
            <span class="n">seq1</span>        <span class="n">GBMAA1117</span><span class="o">-</span><span class="mi">14</span>    <span class="mf">90.6</span>    <span class="mi">265</span>     <span class="mf">84.7</span> <span class="n">Animalia</span><span class="p">;</span><span class="n">Porifera</span><span class="p">;</span><span class="n">Demospongiae</span><span class="p">;</span><span class="n">Haplosclerida</span><span class="p">;</span><span class="n">Phloeodictyidae</span><span class="p">;;</span><span class="n">Calyx</span><span class="p">;</span><span class="n">Calyx</span> <span class="n">podatypa</span>
</pre></div>
</div>
</dd></dl>

</div>
<div class="section" id="otu-annotation">
<span id="annotate-otu"></span><h2>OTU Annotation<a class="headerlink" href="#otu-annotation" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="otu.Annotate_OTU_Table_Command.Annotate_OTU_Table_Command">
<em class="property">class </em><code class="descclassname">otu.Annotate_OTU_Table_Command.</code><code class="descname">Annotate_OTU_Table_Command</code><span class="sig-paren">(</span><em>args_</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/otu/Annotate_OTU_Table_Command.html#Annotate_OTU_Table_Command"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#otu.Annotate_OTU_Table_Command.Annotate_OTU_Table_Command" title="Permalink to this definition">¶</a></dt>
<dd><dl class="docutils">
<dt>Annotates an OTU table with Taxonomic names by replacing sequence names in the OTU table with their identified</dt>
<dd>taxonomies.</dd>
<dt><strong>Inputs</strong>:</dt>
<dd><blockquote class="first">
<div><ul class="simple">
<li>An <a class="reference internal" href="../file_types.html#otu-table"><span class="std std-ref">OTU Table</span></a> to annotate.</li>
<li>One or more <a class="reference internal" href="../file_types.html#tax"><span class="std std-ref">Tax file</span></a> files to read annotations from.</li>
</ul>
</div></blockquote>
<dl class="docutils">
<dt><strong>Outputs</strong>:</dt>
<dd><ul class="first last simple">
<li>An <a class="reference internal" href="../file_types.html#otu-table"><span class="std std-ref">OTU Table</span></a> with sequence names replaced by taxonomic names in the input .tax file.</li>
</ul>
</dd>
<dt><strong>Notes</strong>:</dt>
<dd><ul class="first last simple">
<li>The input annotation file(s) should list only one identification per sequence name.  If you find more than             one taxonomic identity for a sequence, choose only one to include in the input .tax file(s).</li>
</ul>
</dd>
</dl>
<p><strong>Example</strong>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">./</span>
    <span class="n">matrix</span><span class="o">.</span><span class="n">txt</span>
        <span class="n">OTU</span>     <span class="n">Sample1</span> <span class="n">Sample2</span> <span class="n">Sample3</span> <span class="n">Sample4</span>
        <span class="n">seq3</span>    <span class="mi">1</span>       <span class="mi">2</span>       <span class="mi">0</span>       <span class="mi">0</span>
        <span class="n">seq1</span>    <span class="mi">2</span>       <span class="mi">1</span>       <span class="mi">1</span>       <span class="mi">0</span>

    <span class="n">data</span><span class="o">.</span><span class="n">tax</span><span class="p">:</span>
        <span class="n">seq1</span>    <span class="mi">94483305</span>        <span class="mf">99.4</span>    <span class="mi">173</span>     <span class="mf">55.4</span>    <span class="n">Chordata</span><span class="p">:</span><span class="n">Mammalia</span><span class="p">:</span><span class="n">Primates</span><span class="p">:</span><span class="n">Hominidae</span><span class="p">:</span><span class="n">Homo</span><span class="p">:</span><span class="n">Homo</span> <span class="n">sapiens</span>
</pre></div>
</div>
<p><code class="docutils literal"><span class="pre">$</span> <span class="pre">python</span> <span class="pre">chewbacca.py</span> <span class="pre">annotate_matrix</span> <span class="pre">-i</span> <span class="pre">matrix.txt</span> <span class="pre">-a</span> <span class="pre">data.tax</span> <span class="pre">-o</span> <span class="pre">rslt</span></code></p>
<div class="last highlight-default"><div class="highlight"><pre><span></span><span class="n">rslt</span><span class="o">/</span>
    <span class="n">matrix</span><span class="o">.</span><span class="n">txt</span>
        <span class="n">OTU</span>     <span class="n">Sample1</span> <span class="n">Sample2</span> <span class="n">Sample3</span> <span class="n">Sample4</span>
        <span class="n">seq3</span>    <span class="mi">1</span>       <span class="mi">2</span>       <span class="mi">0</span>       <span class="mi">0</span>
        <span class="n">Chordata</span><span class="p">:</span><span class="n">Mammalia</span><span class="p">:</span><span class="n">Primates</span><span class="p">:</span><span class="n">Hominidae</span><span class="p">:</span><span class="n">Homo</span><span class="p">:</span><span class="n">Homo</span> <span class="n">sapiens</span>  <span class="mi">2</span>       <span class="mi">1</span>       <span class="mi">1</span>       <span class="mi">0</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Commands</a><ul>
<li><a class="reference internal" href="#error-correction">Error Correction</a></li>
<li><a class="reference internal" href="#assembling-sequences">Assembling Sequences</a></li>
<li><a class="reference internal" href="#demultiplexing-by-barcode">Demultiplexing by Barcode</a></li>
<li><a class="reference internal" href="#sequence-renaming">Sequence Renaming</a></li>
<li><a class="reference internal" href="#adapter-removal">Adapter Removal</a></li>
<li><a class="reference internal" href="#quality-cleaning">Quality Cleaning</a></li>
<li><a class="reference internal" href="#file-conversion">File Conversion</a></li>
<li><a class="reference internal" href="#dereplication">Dereplication</a></li>
<li><a class="reference internal" href="#file-splitting">File Splitting</a></li>
<li><a class="reference internal" href="#file-merging">File Merging</a></li>
<li><a class="reference internal" href="#file-cleaning">File Cleaning</a></li>
<li><a class="reference internal" href="#deep-cleaning">Deep Cleaning</a></li>
<li><a class="reference internal" href="#deep-cleaning-repair">Deep Cleaning Repair</a></li>
<li><a class="reference internal" href="#sequence-clustering">Sequence Clustering</a></li>
<li><a class="reference internal" href="#otu-table-construction">OTU Table Construction</a></li>
<li><a class="reference internal" href="#otu-identification">OTU Identification</a></li>
<li><a class="reference internal" href="#otu-annotation">OTU Annotation</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/api/commands.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="../index.html">Chewbacca 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, Mahdi Belcaid Greg Burgess.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.5.
    </div>
  </body>
</html>